File: iPhone/VerBank-CSTSv3-IOSClientApi/ClientStation/api/event/temp/API_IDEvent_Spanwser.m
////
////  API_IDEvent_Spanwser.m
////  JEDIv7-CSTSv3-ClientAPI
////
////  Created by baolin on 13/5/15.
////  Copyright (c) 2013年 wangyubo. All rights reserved.
////
//
//#import "API_IDEvent_Spanwser.h"
//
//@implementation API_IDEvent_Spanwser
//
//
//-(id)init{
//    if(self=[super init]){
//        container=[[API_IDEvent_EnquirerContainer alloc]init];
//        eventList=[[NSMutableArray alloc]init];
//    }
//    return self;
//}
//
//-(void)inited{
//    isDestroy=false;
//    [self run];
//}
//
//-(void)destroy{
//    isDestroy=true;
//}
//
//-(void)addListener:(API_IDEventListener *)listener eventName:(NSString *)eventName{
//    [container addListener:listener eventName:eventName];
//}
//
//-(void)removeListener:(API_IDEventListener *)listener eventName:(NSString *)eventName{
//    [container removeListener:listener eventName:eventName];
//}
//
//-(void) removeListener:(API_IDEventListener *)listener{
//    [container removeListener:listener];
//}
//
//-(void)fireEventChanged:(API_IDEvent *)event{
//    @synchronized(eventList)
//    {
//        [eventList addObject:event]; 
//    }
//}
//
//-(void)_notifyEvent:(API_IDEvent *)event
//{
//    NSMutableArray *list=[container getEnquiredListeners:[event getEventName]];
//    for (int i=0; i<[list count]; i++) {
//        API_IDEventListener *listener=[list objectAtIndex:i];
//        @try {
//            [[API_IDEventThreadPool getInstance]notifyEvent:listener event:event];
//        }
//        @catch (NSException *exception) {
//            NSLog(@"Exception %@,%@ ",exception.name,exception.reason);
//        }
//        
//    }
//}
//
//-(void)run{
//    while (!isDestroy) {
//        API_IDEvent *event=nil;
//        @synchronized(eventList){
//            if (eventList ==nil||[eventList count]==0) {
//                event=[eventList objectAtIndex:[eventList count]-1];
//                [eventList removeLastObject];
//            }
//            if(event!=nil){
//                [self _notifyEvent:event];
//            }else{
//                @synchronized(eventList){
//                    @try {
//                        [NSThread sleepForTimeInterval:5];
//                    }
//                    @catch (NSException *exception) {
//                    }
//                }
//            }
//        }
//    }
//}
//@end

File: WeChatService/EX02-CSTSv3-ClientAPI/src/jedi/ex01/client/station/api/event/API_IDEvent_Spanwser.java
package jedi.ex01.client.station.api.event;

import java.util.LinkedList;
import java.util.List;

public class API_IDEvent_Spanwser implements Runnable {
	private boolean isDestroy = false;
	private Thread eventThread = null;
	private API_IDEvent_EnquirerContainer container = new API_IDEvent_EnquirerContainer();

	private LinkedList<API_IDEvent> eventList = new LinkedList<API_IDEvent>();

	public void init() {
		isDestroy = false;
		eventThread = new Thread(this);
		eventThread.start();
	}

	public void destroy() {
		isDestroy = true;
		if (eventThread != null) {
			eventThread.interrupt();
		}
	}

	public void addListener(API_IDEventListener listener, String eventName) {
		container.addListener(listener, eventName);
	}

	public void removeListener(API_IDEventListener listener, String eventName) {
		container.removeListener(listener, eventName);
	}

	public void removeListener(API_IDEventListener listener) {
		container.removeListener(listener);
	}

	public void fireEventChanged(API_IDEvent event) {
		synchronized (eventList) {
			eventList.addFirst(event);
			eventList.notifyAll();
		}
	}

	private void _notifyEvent(API_IDEvent event) {
		List<API_IDEventListener> list = container.getEnquiredListeners(event
				.getEventName());
		for (API_IDEventListener listener : list) {
			try {
//				listener.onStationEvent(event);
				API_IDEventThreadPool.getInstance().notifyEvent(listener, event);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}

	public void run() {
		while (!isDestroy) {
			API_IDEvent event = null;
			synchronized (eventList) {
				if (!eventList.isEmpty()) {
					event = eventList.removeLast();
				}
			}
			if (event != null) {
				_notifyEvent(event);
			} else {
				synchronized (eventList) {
					try {
						eventList.wait(5000);
					} catch (Exception e) {
					}
				}
			}
		}
	}
}
